eventhubs-streamanalytics-eventhubs/tools/eh-asa-perfmon/EventReceiver.cs

```diff
@@ -0,0 +1,119 @@
+
+                        var events = consumerClient.ReadEventsAsync(startReadingAtEarliestEvent: false, readOptions, cancellationToken);
+
+                        await foreach (PartitionEvent currentEvent in events)
+                        {
+                            var eventData = currentEvent.Data;
+
+                            // Because publishing and receiving events is asynchronous, the events that published may not
+                            // be immediately available for our consumer to see, so we'll have to guard against an empty event being sent
+                            if (eventData == null) continue;
+
+                            int eventCount = 0;
+                            var listTimeSpan = new List<TimeSpan>();
+                            var listDateTime = new List<DateTimeOffset>();
+
+                            var eventBody = Encoding.UTF8.GetString(eventData.Body.ToArray()).Split('\n');
+
+                            foreach (var bodyInfo in eventBody)
+                            {
+                                try
+                                {
+                                    var message = JsonSerializer.Deserialize<Dictionary<string, object>>(bodyInfo);
+                                    eventCount++;
+
+                                    var timeCreated = DateTime.Parse(message["createdAt"].ToString());
+                                    var timeIn = DateTime.Parse(message["EventEnqueuedUtcTime"].ToString());
+                                    var timeProcessed = DateTime.Parse(message["EventProcessedUtcTime"].ToString());
+                                    var timeAsaProcessed = DateTime.Parse(message["ASAProcessedUtcTime"].ToString());
+                                    var timeOut = eventData.EnqueuedTime.UtcDateTime.ToLocalTime();
+
+                                    listDateTime.Add(timeIn);
+                                    var elapsed = timeOut - timeIn;
+                                    listTimeSpan.Add(elapsed);
+                                }
+                                catch (Exception ex)
+                                {
+                                    Console.WriteLine("Error while parsing event body.");
+                                    Console.WriteLine("Error:" + ex.Message);
+                                    Console.WriteLine("Message:" + bodyInfo);
+                                }
+                            }
+
+                            var batchFrom = listDateTime.Min();
+                            var batchTo = listDateTime.Min();
+                            var minLatency = listTimeSpan.Min().TotalMilliseconds;
+                            var maxLatency = listTimeSpan.Max().TotalMilliseconds;
+                            var avgLatency = Math.Round(listTimeSpan.Average(ts => ts.TotalMilliseconds), 0);
+
+                            lock (_lock)
```

Comment:
Unless I'm missing something, there's no need to `lock` here.    In the previous version, every partition was independently writing to the `Console`, which would have caused interleaving.   Since this version has only one loop that reads from all partitions, there is only a single thread writing... no need to synchronize.

Reply:
Resolved